### PR TITLE
Fix tslib resolutions

### LINKED ISSUE #57688: Combining `importHelpers: true` and `module: preserve` results in TS2354
### 🔎 Search Terms

TS2354 This syntax requires an imported helper but module 'tslib' cannot be found.
importHelpers
module
preserve

### 🕗 Version & Regression Information

Occurs in 5.4.2 and in ^5.5.0-dev.20240307.

- This changed between versions 5.3.3 and 5.4.2

^ This is not technically true because `"module": "preserve"` wasn't added until 5.4.2.

### ⏯ Playground Link

https://github.com/jgoz/typescript-module-preserve-issue

### 💻 Code

```ts
export class Foo {
  #foo: boolean;

  constructor() {
    this.#foo = false; // This syntax requires an imported helper but module 'tslib' cannot be found. ts(2354)
  }

  foo(): boolean {
    return this.#foo;
  }
}
```

```json
{
  "compilerOptions": {
    "moduleResolution": "bundler",
    "module": "preserve",
    "importHelpers": true,
    "outDir": "dist",
    "target": "es2021"
  }
}
```

### 🙁 Actual behavior

Error message indicates that tslib is not present, but it is.

### 🙂 Expected behavior

No error.

### Additional information about the issue

Not sure if this is a TS issue or an issue with the export maps in `tslib`.

### LINKED ISSUE #58450: `importHelpers` with CommonJS `.js` file results in "This syntax requires an imported helper but module 'tslib' cannot be found"
### 🔎 Search Terms

tslib javascript

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about quantum tunneling

### ⏯ Playground Link

_No response_

### 💻 Code

```ts
// index.js
module.exports = function foo(args) {
  const { bar, ...extraArgs } = args;
  return extraArgs;
}
```

```sh
npm i tslib
tsc --checkJs --importHelpers --module commonjs --noEmit index.js
```

### 🙁 Actual behavior

```
index.js:2:19 - error TS2354: This syntax requires an imported helper but module 'tslib' cannot be found.

2   const { bar, ...extraArgs } = args;
                    ~~~~~~~~~
```

### 🙂 Expected behavior

No error

### Additional information about the issue

_No response_

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #57688
Also discovered and fixed #58450 in the process

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
idk which perf tests use tslib, but

@typescript-bot perf test this

**Comment by jakebailey** [comment]
I'm not sure any of them do, actually...

**Comment by andrewbranch** [comment]
@typescript-bot test top400

File: src/compiler/program.ts [hunk 1]
```diff
@@ -3390,11 +3390,11 @@ export function createProgram(rootNamesOrOptions: readonly string[] | CreateProg
...
-        // helpers library.
-        if (
-            (getIsolatedModules(options) || isExternalModuleFile)
-            && !file.isDeclarationFile
-        ) {
+        // helpers library. (A JavaScript file without `externalModuleIndicator` set might be
+        // a CommonJS module; `commonJsModuleIndicator` doesn't get set until the binder has
+        // run. We synthesize a helpers import for it just in case; it will never be used if
+        // the binder doesn't find and set a `commonJsModuleIndicator`.)
+        if (isJavaScriptFile || (getIsolatedModules(options) || isExternalModule(file)) && !file.isDeclarationFile) {
```
Comment (sheetalkamat) [inline src/compiler/program.ts:3397]:
```suggestion
        if (isJavaScriptFile || (!file.isDeclarationFile && (getIsolatedModules(options) || isExternalModule(file)))) {
```

**Review by sheetalkamat — COMMENTED** [review]
looks good

**Pull request merged**